File: backend/app/schemas/opportunity.py
from datetime import datetime

from app.schemas.base import CamelModel
from app.schemas.bom import DefectiveComponentResponse, RepairSkillSummaryResponse
from app.schemas.catalog_product import CatalogProductDetailResponse, CatalogProductSummaryResponse
from app.schemas.common import MarketplaceSource, UserDealStatus
from app.schemas.listing import ListingDetailResponse
from app.schemas.market_comp import MarketCompSummaryResponse


class DealScoreResponse(CamelModel):
    total: int
    restorer_score: int
    harvester_score: int
    valuation_confidence: float


class OpportunityFinancialsResponse(CamelModel):
    purchase_price: float
    shipping_cost: float
    total_replacement_cost: float
    total_acquisition_and_repair_cost: float
    estimated_working_market_value: float
    projected_restorer_net: float
    projected_harvest_yield: float
    roi_percent: float


class OpportunitySummaryResponse(CamelModel):
    listing_id: str
    source: MarketplaceSource
    title: str
    description: str | None = None
    url: str
    price: float
    shipping_cost: float
    image_url: str | None = None
    condition: str
    listed_at: datetime | None = None
    product: CatalogProductSummaryResponse | None = None
    required_repair_skills: list[str]
    projected_restorer_net: float
    projected_harvest_yield: float
    estimated_repair_cost: float
    deal_score: int
    user_deal_status: UserDealStatus | None = None


class OpportunityListResponse(CamelModel):
    items: list[OpportunitySummaryResponse]
    page: int
    page_size: int
    total: int


class UserDealSummaryResponse(CamelModel):
    id: str
    listing_id: str | None = None
    status: UserDealStatus
    notes: str | None = None
    created_at: datetime
    updated_at: datetime | None = None


class OpportunityDetailResponse(CamelModel):
    listing: ListingDetailResponse
    product: CatalogProductDetailResponse | None = None
    defects: list[DefectiveComponentResponse]
    required_skills: list[RepairSkillSummaryResponse]
    market_comps: list[MarketCompSummaryResponse]
    financials: OpportunityFinancialsResponse
    deal_score: DealScoreResponse
    user_deal: UserDealSummaryResponse | None = None

File: frontend/src/types/opportunity.ts
import type { CatalogProductDetail, CatalogProductSummary, DefectiveComponent, MarketCompSummary, RepairSkillSummary } from "./catalog";
import type { UserDealStatus, UserDealSummary } from "./deal";
import type { ListingDetail } from "./listing";

export type MarketplaceSource = "ebay" | "facebook_marketplace";

export type { UserDealStatus };

export type OpportunitySort = "score_desc" | "profit_desc" | "roi_desc" | "price_asc" | "newest";

export interface RadarFilters {
  query?: string;
  source?: MarketplaceSource;
  category?: string;
  minProfit?: number;
  minRoi?: number;
  maxPrice?: number;
  condition?: string;
  requiredSkill?: string;
  persona?: string;
  sort?: OpportunitySort;
  page?: number;
  pageSize?: number;
}

export interface DealScore {
  total: number;
  restorerScore: number;
  harvesterScore: number;
  valuationConfidence: number;
}

export interface OpportunityFinancials {
  purchasePrice: number;
  shippingCost: number;
  totalReplacementCost: number;
  totalAcquisitionAndRepairCost: number;
  estimatedWorkingMarketValue: number;
  projectedRestorerNet: number;
  projectedHarvestYield: number;
  roiPercent: number;
}

export interface OpportunitySummary {
  listingId: string;
  source: MarketplaceSource;
  title: string;
  description?: string;
  url: string;
  price: number;
  shippingCost: number;
  imageUrl?: string;
  condition: string;
  listedAt?: string;
  product?: CatalogProductSummary;
  requiredRepairSkills: string[];
  projectedRestorerNet: number;
  projectedHarvestYield: number;
  estimatedRepairCost: number;
  dealScore: number;
  userDealStatus: UserDealStatus | null;
}

export interface OpportunityListResponse {
  items: OpportunitySummary[];
  page: number;
  pageSize: number;
  total: number;
}

export interface OpportunityDetail {
  listing: ListingDetail;
  product: CatalogProductDetail | null;
  defects: DefectiveComponent[];
  requiredSkills: RepairSkillSummary[];
  marketComps: MarketCompSummary[];
  financials: OpportunityFinancials;
  dealScore: DealScore;
  userDeal: UserDealSummary | null;
}
